File: depth_first_search/count-paths-for-a-sum.js
class TreeNode {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

function find_paths(root, S) {
    function dfs(node, path = []) {
        if (!node) return 0;
        path.push(node.value);
        let [path_sum, total] = [0, 0];
        for (let i = path.length - 1; i >= 0; i--) {
            path_sum += path[i];
            if (path_sum === S) total++;
        }
        total += dfs(node.left, path) + dfs(node.right, path)
        path.pop();
        return total;
    }

    return dfs(root, [])
}

root = new TreeNode(5)
root.left = new TreeNode(7)
root.right = new TreeNode(1)
root.left.left = new TreeNode(4)
root.right.left = new TreeNode(10)
root.right.right = new TreeNode(5)
root.right.left.left = new TreeNode(1)
console.log(`Tree has paths: ${find_paths(root, 11)}`)

File: depth_first_search/python/count-paths-for-a-sum.py
class TreeNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def find_paths(root, S):
    def dfs(node, path) -> int:
        if node is None:
            return 0

        path.append(node.value)
        path_sum = total = 0

        for i in range(len(path) - 1, -1, -1):
            path_sum += path[i]
            if path_sum == S:
                total += 1

        total += dfs(node.left, path) + dfs(node.right, path)

        path.pop()

        return total

    return dfs(root, [])


if __name__ == '__main__':
    root = TreeNode(5)
    root.left = TreeNode(7)
    root.right = TreeNode(1)
    root.left.left = TreeNode(4)
    root.right.left = TreeNode(10)
    root.right.right = TreeNode(5)
    root.right.left.left = TreeNode(1)
    print(f"Tree has paths: {find_paths(root, 11)}")
